Add language picker

wold5 opened the pull request:
Hi Jens, 

I've done a global check of your feedback and changes, re-applied them to the 1.3.1, and rebased to include the latest commits. The [old pull request](https://github.com/bookalope/InDesign-CEP/pull/17) has been revoked.

1. Saw you removed commits re handling token field changes (authenticationChanged, the blocking (skeleton) function and keypress eventListeners). But those can be added back in later.

2. Saw you simplified parts of the spectrum picker code (= better).

3. One remark about the eventlistener that handles closeAllDropdowns() when clicking outside a picker/dropdown menu

```
                // Clicking outside of the Dropdown or its popover closes any open popover.
                document.addEventListener("click", function (event) {
                    if (!isAncestorOf(event.target, ".dropdown-select")) {
                        closeAllDropdowns();
                    }
                });
```
As it attaches to `document`, it's re-added on each invocation of refreshSpectrumDropdowns().
It's OK if that occurs just once, otherwise it's best added to the 'main' function (or implemented differently?). Scoping may be an issue, if only from a sense of keeping things orderly, but I didn't experience any issues with my edit.

wold5 committed:
apply 351d8bb-add-language-server-support.diff

- implements final (working) version of getLanguages (Jens, based off
original/Wold5)
- renames variable language to locale (Jens)

wold5 committed:
apply d71fec0-populate-language-picker.diff

- add refresh panels function (Jens, based off wold5)
- move adobe picker code into separate file (Jens, reworking/reverting
wold5 changes)

wold5 committed:
apply 7bc2b1d-improve-token-handling.diff

- Added token checks (Jens)
- Added UI eventlisteners (wold5)

jenstroeger commented:
Thank you @wold5 for the update!

> I've done a global check of your feedback and changes, re-applied them to the 1.3.1, and rebased to include the latest commits.

So this PR is based on current master commit https://github.com/bookalope/InDesign-CEP/commit/f7468f8565453bf5bed283d8ab1d51129e5992ad, correct?

> The [old pull request](https://github.com/bookalope/InDesign-CEP/pull/17) has been revoked.

I think we can close #17 then, retitle this to “Populate language picker”, and change it from Draft to open.

> 1. Saw you removed commits re handling token field changes (authenticationChanged, the blocking (skeleton) function and keypress eventListeners). But those can be added back in later.

The handling of the Token input field and Beta checkbox isn’t as elaborate as in #17 and many scenarios are handled [on `change`](https://github.com/wold5/InDesign-CEP/blob/4cbab85e09a102c889353cefb5f2a4088e3182b1/extensions/Bookalope/js/bookalope.js#L1071-L1106) — it would be a good idea to also handle [`blur`](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event) at least? However, the [`keypress` event](https://developer.mozilla.org/en-US/docs/Web/API/Element/keypress_event) is deprecated and replaced by the [`keydown` event](https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event), and we can add support for it later if there’s confusion when changing & typing a token?

> 2. Saw you simplified parts of the spectrum picker code (= better).

Tried to 🤓

> 3. One remark about the eventlistener that handles closeAllDropdowns() when clicking outside a picker/dropdown menu
> 
> ```javascript
>                 // Clicking outside of the Dropdown or its popover closes any open popover.
>                 document.addEventListener("click", function (event) {
>                     if (!isAncestorOf(event.target, ".dropdown-select")) {
>                         closeAllDropdowns();
>                     }
>                 });
> ```
> 
> As it attaches to `document`, it's re-added on each invocation of `refreshSpectrumDropdowns()`. It's OK if that occurs just once, otherwise it's best added to the 'main' function (or implemented differently?). Scoping may be an issue, if only from a sense of keeping things orderly, but I didn't experience any issues with my edit.

Hmm, I think that implementation might be ok: quote from [the documentation](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) —

> If the function or object is already in the list of event listeners for this target, the function or object is not added a second time.

I think that means that although the `document.addEventListener()` is called multiple times, the handler function itself is added only once and subsequent calls have no effect. What do you think?

wold5 commented:
> Thank you @wold5 for the update!
> 
> > I've done a global check of your feedback and changes, re-applied them to the 1.3.1, and rebased to include the latest commits.
> 
> So this PR is based on current master commit [f7468f8](https://github.com/bookalope/InDesign-CEP/commit/f7468f8565453bf5bed283d8ab1d51129e5992ad), correct?

Yes. BTW, if it becomes too much of a hassle, feel free to re-use your own commits, and perhaps reference this PR ;)

> 
> > The [old pull request](https://github.com/bookalope/InDesign-CEP/pull/17) has been revoked.
> 
> I think we can close #17 then, retitle this to “Populate language picker”, and change it from Draft to open.

Done. Hadn't expected you could already see the draft! So this is a quick reply.

> 
> > 1. Saw you removed commits re handling token field changes (authenticationChanged, the blocking (skeleton) function and keypress eventListeners). But those can be added back in later.
> 
> The handling of the Token input field and Beta checkbox isn’t as elaborate as in #17 and many scenarios are handled [on `change`](https://github.com/wold5/InDesign-CEP/blob/4cbab85e09a102c889353cefb5f2a4088e3182b1/extensions/Bookalope/js/bookalope.js#L1071-L1106) — it would be a good idea to also handle [`blur`](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event) at least? However, the [`keypress` event](https://developer.mozilla.org/en-US/docs/Web/API/Element/keypress_event) is deprecated and replaced by the [`keydown` event](https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event), and we can add support for it later if there’s confusion when changing & typing a token?
> 

I'll get back on this, next Sat/Sunday.

> > 3. One remark about the eventlistener that handles closeAllDropdowns() when clicking outside a picker/dropdown menu

> > As it attaches to `document`, it's re-added on each invocation of `refreshSpectrumDropdowns()`. It's OK if that occurs just once, otherwise it's best added to the 'main' function (or implemented differently?). Scoping may be an issue, if only from a sense of keeping things orderly, but I didn't experience any issues with my edit.
> 
> Hmm, I think that implementation might be ok: quote from [the documentation](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) —
> 
> > If the function or object is already in the list of event listeners for this target, the function or object is not added a second time.
> 
> I think that means that although the `document.addEventListener()` is called multiple times, the handler function itself is added only once and subsequent calls have no effect. What do you think?

Aren't anonymous functions the exception? From the same documentation page:

> Note: If a particular anonymous function is in the list of event listeners registered for a certain target, and then later in the code, an identical anonymous function is given in an addEventListener call, the second function will also be added to the list of event listeners for that target.
> 
> Indeed, anonymous functions are not identical even if defined using the same unchanging source-code called repeatedly, even if in a loop. 

One way is checking the event handlers using the Developer Mode in debug mode. IIRC, more than one listener was added, but it's been a while, and it might have involved another project.

jenstroeger commented:
> Yes. BTW, if it becomes too much of a hassle, feel free to re-use your own commits, and perhaps reference this PR ;)

All good, this is your contribution 👍🏼

> I'll get back on this, next Sat/Sunday.

The `blur`? Great, thank you!

> Aren't anonymous functions the exception? From the same documentation page:
> 
> > Note: If a particular anonymous function is in the list of event listeners registered for a certain target, and then later in the code, an identical anonymous function is given in an addEventListener call, the second function will also be added to the list of event listeners for that target.
> > Indeed, anonymous functions are not identical even if defined using the same unchanging source-code called repeatedly, even if in a loop.
> 
> One way is checking the event handlers using the Developer Mode in debug mode. IIRC, more than one listener was added, but it's been a while, and it might have involved another project.

Gah, I was worried that this may build a list of that same function (should’ve [RTFM](https://en.wikipedia.org/wiki/RTFM)’ed the whole page). Happy to hoist the anonymous function into a proper named one…

wold5 commented:
> it would be a good idea to also handle blur at least?

Mmhm, "change" seems to operates like blur*, so it would suffice to refresh the UI from the `input-bookalope-token` eventhandler, as the "input-bookalope-beta" handler does (line 1099). Is that OK with you? Else I'll need to think of something more elaborate.

_* "change" for input elements triggers after a) loosing focus _and_ b) a value change. The exception with (b) is the beta toggle, but that's handled independently. Inserting programmatically would fail (a), but can be manually triggered._

> However, the [keypress event](https://developer.mozilla.org/en-US/docs/Web/API/Element/keypress_event) is deprecated and replaced by the [keydown event](https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event), and we can add support for it later if there’s confusion when changing & typing a token?

Agreed, it can wait. Indeed, I found the out-of-focus behavior not too intuitive. Thinking about it, a mock "load" button might suffice to force that condition. Positioning and styling with Spectrum is a bit iffy to get right, but the following should work (inserted below the text input tag):
```
				<button id="button-mock-reload" class="spectrum-Button spectrum-Button--fill spectrum-Button--secondary spectrum-Button--sizeS">
					<span class="spectrum-Button-label">Reload</span>
				</button>
```

jenstroeger commented:
> Mmhm, "change" seems to operates like blur*, so it would suffice to refresh the UI from the `input-bookalope-token` eventhandler, as the "input-bookalope-beta" handler does (line 1099). Is that OK with you? Else I'll need to think of something more elaborate.

Sounds good 👍🏼 

> Agreed, it can wait. Indeed, I found the out-of-focus behavior not too intuitive. Thinking about it, a mock "load" button might suffice to force that condition. Positioning and styling with Spectrum is a bit iffy to get right, but the following should work (inserted below the text input tag):
> 
> ```html
> <button id="button-mock-reload" class="spectrum-Button spectrum-Button--fill spectrum-Button--secondary spectrum-Button--sizeS">
>   <span class="spectrum-Button-label">Reload</span>
> </button>
> ```

You mean to explicitly reload server-side data when either the token textfield or beta checkbox has changed — in addition to the `"change"` event handler? (Kinda makes me wish that Spectrum had [input groups like Bootstrap](https://getbootstrap.com/docs/5.2/forms/input-group/).)

wold5 commented:
> > Agreed, it can wait. Indeed, I found the out-of-focus behavior not too intuitive. Thinking about it, a mock "load" button might suffice to force that condition. Positioning and styling with Spectrum is a bit iffy to get right, but the following should work (inserted below the text input tag):
> > ```
> > <button id="button-mock-reload" class="spectrum-Button spectrum-Button--fill spectrum-Button--secondary spectrum-Button--sizeS">
> >   <span class="spectrum-Button-label">Reload</span>
> > </button>
> > ```
> 
> You mean to explicitly reload server-side data when either the token textfield or beta checkbox has changed — in addition to the `"change"` event handler? (Kinda makes me wish that Spectrum had [input groups like Bootstrap](https://getbootstrap.com/docs/5.2/forms/input-group/).)

Not necessarily in addition to: the button needn't do anything, except that clicking it changes the focus, making the "change" event happen (for the input field) -- provided the input value _also_ changed --  (which in turn reloads server-side data).

Using just the button's "click" event would suffice also (and is more logical...). And simpler than the keydown/timer combo (Bootstrap's buttons are great BTW.).

jenstroeger commented:
> Not necessarily in addition to: the button needn't do anything, except that clicking it changes the focus, making the "change" event happen (for the input field) -- provided the input value _also_ changed -- (which in turn reloads server-side data).
> 
> Using just the button's "click" event would suffice also (and is more logical...). And simpler than the keydown/timer combo.

You know, I’m tempted to vote for a simple `"change"` event. Most users these days probably know that tabbing or entering out of the textfield is necessary to commit its value. A simple click outside (i.e. `"blur"`) [also counts as textfield change](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event) so we should be good 🤔

> (Bootstrap's buttons are great BTW.)

I’m a big fan 🤓

wold5 commented:
> > Not necessarily in addition to: the button needn't do anything, except that clicking it changes the focus, making the "change" event happen (for the input field) -- provided the input value _also_ changed -- (which in turn reloads server-side data).
> > Using just the button's "click" event would suffice also (and is more logical...). And simpler than the keydown/timer combo.
> 
> You know, I’m tempted to vote for a simple `"change"` event. Most users these days probably know that tabbing or entering out of the textfield is necessary to commit its value. A simple click outside (i.e. `"blur"`) [also counts as textfield change](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event) so we should be good 🤔

Then let's keep things simple 😃 

Question, can refreshPanelsFromServerConfiguration() handle an empty (=== "") token? Will it contact the server, throw and catch, thus defaulting to the en-US locale?

On a related note, #17 includes a (stub) function for blocking UI elements. It's complicated, requiring adding Spectrum "is-disabled" classes to items, so I never fully implemented it. If you do want that, let me know.

jenstroeger commented:
> Then let's keep things simple 😃

Okies 👍🏼

> Question, can refreshPanelsFromServerConfiguration() handle an empty (=== "") token? Will it contact the server, throw and catch, thus defaulting to the en-US locale?

Funny you ask, that’s commit https://github.com/bookalope/InDesign-CEP/commit/7400730b5877c6813909e045e4ae6e0e37feada3 a week ago. The change in that commit prevents a request with invalid token being sent to the server:

https://github.com/bookalope/InDesign-CEP/blob/f7468f8565453bf5bed283d8ab1d51129e5992ad/extensions/Bookalope/js/uses/BookalopeClient.js#L93-L96

I guess
```javascript
if (!token || !isToken(token)) { }
```
would also suffice 🤓

So, you’re correct: the `bookalope.getLanguages()` Promise will fail and resolve to using the default `en-US` locale ([code](https://github.com/wold5/InDesign-CEP/blob/4cbab85e09a102c889353cefb5f2a4088e3182b1/extensions/Bookalope/js/bookalope.js#L869-L874)).

> On a related note, #17 includes a (stub) function for blocking UI elements. It's complicated, requiring adding Spectrum "is-disabled" classes to items, so I never fully implemented it. If you do want that, let me know.

I’ve not dived into Spectrum enough yet to find their recommendation on disabling multiple fields. In other projects I’ve always tried to wrap multiple input fields into a `<fieldset>` because [setting the `disabled` attribute on that set disables all descendent fields](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset#attr-disabled). For this extension’s panel, I’ve not played around with these different options…

wold5 committed:
reload server-side data on token change

wold5 commented:
> > Question, can refreshPanelsFromServerConfiguration() handle an empty (=== "") token? Will it contact the server, throw and catch, thus defaulting to the en-US locale?
> 
> Funny you ask, that’s commit [7400730](https://github.com/bookalope/InDesign-CEP/commit/7400730b5877c6813909e045e4ae6e0e37feada3) a week ago. The change in that commit prevents a request with invalid token being sent to the server:
> 
> https://github.com/bookalope/InDesign-CEP/blob/f7468f8565453bf5bed283d8ab1d51129e5992ad/extensions/Bookalope/js/uses/BookalopeClient.js#L93-L96

Thanks for confirming, then it's all OK. Otherwise an additional check was needed.  

The code for refreshing the server-side data has been added. It's a re-application of the beta toggle code.
I tested it more properly now (due to hardware/logistics issues) and it worked. 
A default language is/was not being set, but that's better, thinking about it.

> I guess
> 
> ```js
> if (!token || !isToken(token)) { }
> ```
> 
> would also suffice 🤓

One of those JS things :) (each language has them)

>> On a related note, https://github.com/bookalope/InDesign-CEP/pull/17 includes a (stub) function for blocking UI elements. It's complicated, requiring adding Spectrum "is-disabled" classes to items, so I never fully implemented it. If you do want that, let me know.
> I’ve not dived into Spectrum enough yet to find their recommendation on disabling multiple fields. In other projects I’ve always tried to wrap multiple input fields into a `<fieldset>` because [setting the disabled attribute on that set disables all descendent fields](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset#attr-disabled). For this extension’s panel, I’ve not played around with these different options…

The `<Fieldset>` can toggle the `disabled` attributes. The [Spectrum docs do show the parent `<div>` also needing addition/removal of the class "is-disabled"](https://opensource.adobe.com/spectrum-css/textfield.html#disabled). IIRC, that affects the styling, especially for the picker/dropdown, and less for the text fields and buttons (as those utilize regular HTML tags).

jenstroeger commented on extensions/Bookalope/js/uses/Spectrum.js:
```diff
@@ -0,0 +1,303 @@
+ *
+ *   https://opensource.adobe.com/spectrum-css/picker.html
+ *   https://spectrum.adobe.com/page/picker/
+ *
```
Perhaps we should add this one link to justify the entirety of the JavaScript file 🤓
```
https://github.com/adobe/spectrum-css#where-is-the-javascript
```

jenstroeger reviewed (approved):
Looks great, @wold5 👍🏼
